### ISSUE TITLE
[ZA] Import 2019 election candidates

### ISSUE THREAD

**Issue by chrismytton** [mysociety/pombola#2595]
Need to load the 2019 election candidates into the People's Assembly database.

OpenUp were going to do this, but they're having trouble getting pombola running locally 😞 So we're going to do this instead.

This is the [email I sent to JD](https://groups.google.com/a/mysociety.org/d/msg/pombola/3_yaVu2hocU/OQ-zQ4iaCwAJ) about how we might go about doing this.

>I've just had a look at what we did in 2014 and I think this should be reasonably straightforward.
>
>Everything for the elections in 2014 was done in the pa.org.za app. I think you're already aware, but just a reminder that pa.org.za runs from the pombola codebase. All development is done through GitHub.
>
>**START TECHNICAL BIT**
>
>You can see the format for the 2014 candidate data in pombola/south_africa/data/candidates-2014-elections.csv. Looking at the XLSX files from the page that Gaile linked to in her email it seems that the data for the 2019 elections is in a similar format, with the addition of an IDNumber column, so we should be able to use the code from last time with a few tweaks.
>
>The script we used to import the CSV files into Pombola is pombola/south_africa/management/commands/south_africa_import_election_candidates.py. There are a few "TODO" comments in that script about 2014 dates being hardcoded. There's also code for dealing with ending memberships, which might not be necessary as I bulk-added a 2019 end-date to all MPs a couple of weeks ago.
>
>We can either make some changes to that script to make it work for 2019 (and beyond!), or we can copy the script and make the necessary changes without worrying about backwards compatibility. I think it's probably simpler to copy the script (e.g. to pombola/south_africa/management/commands/south_africa_import_election_candidates_2019.py), then make the necessary changes to the copy for it to work for the 2019 elections. In this case I'm OK with a bit of duplication if it means the individual script can be simpler :)
>
>In theory once the candidates have been imported into the database we might get a nice interface for them at /election/2019/, but in practice we'll probably need to tweak it a bit so things look right. The Django views for the elections are in pombola/south_africa/views/elections.py, and the templates are in pombola/south_africa/templates/south_africa/election. Those might need to be changed once the candidate data is in, but we'll need to ensure they continue to work for the 2014 election pages.
>
>**END TECHNICAL BIT**
>
>I'm afraid I wasn't actually involved in the 2014 elections work myself, so I'm not sure how much more information I'd be able to impart on a call. Hopefully the above is enough to point you in the right direction.
>
>I gleaned most of this information from reading though pull request #1408 on GitHub, which from what I can tell was the main (only?) pull request for the 2014 elections work we did at our end. It might be worth you having a look through that to get some extra context.

**Comment by chrismytton** [comment]
Regarding the import of candidates into the database, I've read through the [import script from 2014](https://github.com/mysociety/pombola/blob/c7d68e692c432ecd99e33c00d3a29225eaa9a49f/pombola/south_africa/management/commands/south_africa_import_election_candidates.py) and I think the following needs doing to get the data into pombola.

- [ ] Convert [candidate data](https://www.elections.org.za/IECOnline/2019-Candidate-Lists) to CSV format and add to this repo
- [ ] Check if the [party mapping file](https://github.com/mysociety/pombola/blob/c7d68e692c432ecd99e33c00d3a29225eaa9a49f/pombola/south_africa/data/parties-2014-elections.csv) from 2014 can be re-used and if there are any additions needed
- Updates to the script
  - [ ] Fix script so it works with the new format CSV, which has an extra column and headers.
  - [ ] Add an election date variable to the import script (either as an argument or just hard-coded) rather than duplicating dates throughout the script
  - [ ] Eliminate all hard-coded 2014 dates
  - [ ] Add variables with hard-coded paths to the data CSVs
  - [ ] Remove code to handle ending memberships, this has already been done manually.
  - [ ] Check that the person search is returning expected results
  - [ ] Check unmatched people by hand and manually match if necessary

Once that is done and the data is in the database then we can look at the front-end and see 1) if it works and 2) what changes need making, if any.

**Comment by chrismytton** [comment]
Have done the initial import using the 2019 script added in #2597, however it seems that the slug generation for people is causing issues as it's not converting the slug to ASCII, so there are unwanted unicode characters in some of them. This then has a knock on effect that once we convert the slugs to ASCII they are actually duplicates of existing people 🤦‍♂ so I'm going through manually and tidying them up.

**Comment by chrismytton** [comment]
Election candidates are now live at https://www.pa.org.za/election/2019/. I'll open a separate ticket for importing the results into the database.
